File: solutions/go/03/dp_paths/main.go
package main

import "fmt"

func paths(h, w int, blocked [][]bool) int {
	dp := make([][]int, h)
	for i := range dp {
		dp[i] = make([]int, w)
	}
	if !blocked[0][0] {
		dp[0][0] = 1
	}
	for i := 0; i < h; i++ {
		for j := 0; j < w; j++ {
			if blocked[i][j] {
				dp[i][j] = 0
				continue
			}
			if i > 0 {
				dp[i][j] += dp[i-1][j]
			}
			if j > 0 {
				dp[i][j] += dp[i][j-1]
			}
		}
	}
	return dp[h-1][w-1]
}

func main() {
	blocked := make([][]bool, 3)
	for i := range blocked {
		blocked[i] = make([]bool, 3)
	}
	blocked[1][1] = true
	fmt.Println(paths(3, 3, blocked))
}

File: solutions/python/03/dp_paths/main.py
def paths(h: int, w: int, blocked: list) -> int:
    dp = [[0] * w for _ in range(h)]
    if not blocked[0][0]:
        dp[0][0] = 1
    for i in range(h):
        for j in range(w):
            if blocked[i][j]:
                dp[i][j] = 0
                continue
            if i:
                dp[i][j] += dp[i - 1][j]
            if j:
                dp[i][j] += dp[i][j - 1]
    return dp[h - 1][w - 1]


def main() -> None:
    blocked = [[False] * 3 for _ in range(3)]
    blocked[1][1] = True
    print(paths(3, 3, blocked))


if __name__ == "__main__":
    main()
